src/InteractiveWindow/Editor/InteractiveWindow_UIThread.cs

```diff
@@ -309,11 +326,134 @@ public void ClearView()
+
+                        case State.WaitingForInput:
+                            State = State.WaitingForInputAndReadingStandardInput;
+                            break;
+                        case State.ExecutingInput:
+                            State = State.ExecutingInputAndReadingStandardInput;
+                            break;
+                        case State.Resetting:
+                            State = State.ResettingAndReadingStandardInput;
+                            break;
+
+                        default:
+                            throw ExceptionUtilities.UnexpectedValue(State);
+                    }
+
+                    Debug.Assert(ReadingStandardInput);
+
+                    _window._buffer.Flush();
+
+                    if (State == State.WaitingForInputAndReadingStandardInput)
+                    {
+                        var snapshot = _window._projectionBuffer.CurrentSnapshot;
+                        var spanCount = snapshot.SpanCount;
+                        if (spanCount > 0 && _window.GetSpanKind(snapshot.GetSourceSpan(spanCount - 1)) == ReplSpanKind.Language)
+                        {
+                            // we need to remove our input prompt.
+                            RemoveLastInputPrompt();
+                        }
+                    }
+
+                    AddStandardInputSpan();
+
+                    _window.Caret.EnsureVisible();
+                    ResetCursor();
+
+                    UncommittedInput = null;
+                    _standardInputStart = _window._standardInputBuffer.CurrentSnapshot.Length;
+
+                    var value = await GetStandardInputValue().ConfigureAwait(true);
+                    Debug.Assert(_window.OnUIThread()); // ConfigureAwait should bring us back to the UI thread.
+                    return value.HasValue
+                        ? new StringReader(value.GetValueOrDefault().GetText())
+                        : null;
+                }
+                catch (Exception e) when (_window.ReportAndPropagateException(e))
+                {
+                    throw ExceptionUtilities.Unreachable;
+                }
+            }
+
             private void CancelStandardInput()
             {
+                SetStandardInputValue(null);
+            }
+
+            private void SetStandardInputValue(SnapshotSpan? value)
+            {
+                _standardInputValue = value;
+                _standardInputValueGuard.Release();
+            }
+
+            private async Task<SnapshotSpan?> GetStandardInputValue()
+            {
+                try
+                {
+                    await _standardInputValueGuard.WaitAsync().ConfigureAwait(true);
+                    Debug.Assert(_window.OnUIThread()); // ConfigureAwait should bring us back to the UI thread.
+                    return _standardInputValue;
+                }
+                catch (Exception e) when (_window.ReportAndPropagateException(e))
```

Comment:
In that case, could this method be inlined in the one caller so there's one less exception handler?